astropy/units/decorators.py

```diff
@@ -127,7 +139,11 @@ def wrapper(*func_args, **func_kwargs):
 
             # Call the original function with any equivalencies in force.
             with add_enabled_equivalencies(self.equivalencies):
-                return wrapped_function(*func_args, **func_kwargs)
+                return_ = wrapped_function(*func_args, **func_kwargs)
+            if isinstance(wrapped_signature.return_annotation, (Unit, Quantity)):
```

Comment:
@astrofrog @mhvk I changed this to verify Unit or Quantity so that `quantity_input` doesn't just eat all the possible uses for return annotations. Opinions?